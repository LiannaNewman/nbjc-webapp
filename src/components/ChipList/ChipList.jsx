import React, { useState, useLayoutEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import { useInView } from 'react-intersection-observer';
import { withStyles } from '@material-ui/core/styles';
import { Chip, Typography } from '@material-ui/core';

const styles = {
  chipWrapper: {
    maxHeight: '100%',
    overflow: 'hidden',
  },
  chip: {
    marginBottom: 8,
    marginRight: 4,
    backgroundColor: '#F2F2F2',
    color: '#666666',
    borderRadius: '4px',
  },
};

const ChipList = ({ chips, classes }) => {
  const parentRef = useRef();
  const [moreChips, setMoreChips] = useState(0);
  const inViewList = [];
  useLayoutEffect(() => {
    // filter out those which are false and count their length
    setMoreChips(inViewList.filter((i) => !i).length);
  }, [inViewList]);

  const chipRenderer = () => (
    chips.map((chip) => {
      const [ref, inView] = useInView({ threshold: 1 });
      inViewList.push(inView);
      return (
        <Chip
          color="primary"
          label={
            <Typography variant="body2">{chip.name}</Typography>
          }
          key={chip.name.replace(/\w/, '')}
          className={classes.chip}
          ref={ref}
          size="small"
        />
      );
    })
  );
  return (
    <>
      <div className={classes.chipWrapper} ref={parentRef}>
        {chipRenderer()}
      </div>
      {moreChips > 0 && (
        <Chip
          color="primary"
          variant="outlined"
          label={`+${moreChips} More`}
        />
      )}
    </>
  );
};

ChipList.propTypes = {
  chips: PropTypes.arrayOf(PropTypes.shape({
    name: PropTypes.string.isRequired,
  })),
  classes: PropTypes.shape({}).isRequired,
};

ChipList.defaultProps = {
  chips: [],
};

export default withStyles(styles)(ChipList);
