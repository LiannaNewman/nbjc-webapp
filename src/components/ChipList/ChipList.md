## ChipList

```jsx
const chips = [{
    name: 'Black Friendly',
  }, {
    name: 'Inclusive',
  }, {
    name: 'Black Owned',
  }, {
    name: 'Gender Neutral Restrooms',
  }, {
    name: 'Accessible',
  }, {
    name: 'Queer hangout space',
  }, {
    name: 'Trans friendly',
  }, {
    name: 'Queer owned',
  }];
<ChipList chips={chips}/>
```