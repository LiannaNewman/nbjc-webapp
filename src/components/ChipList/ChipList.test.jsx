import React from 'react';
import { render } from '@testing-library/react';
import ChipList from './ChipList';

jest.mock('react-intersection-observer', () => ({
  useInView: () => [{}, true],
}));

describe('ChipList', () => {
  test('render', () => {
    const chips = [{
      name: 'Black Friendly',
    }, {
      name: 'Inclusive',
    }, {
      name: 'Black Owned',
    }, {
      name: 'Gender Neutral Restrooms',
    }, {
      name: 'Accessible',
    }, {
      name: 'Queer hangout space',
    }, {
      name: 'Trans friendly',
    }, {
      name: 'Queer owned',
    }];
    const { asFragment } = render(<ChipList chips={chips} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
