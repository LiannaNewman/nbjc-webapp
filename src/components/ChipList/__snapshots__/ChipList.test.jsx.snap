// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChipList render 1`] = `
<DocumentFragment>
  <div
    class="ChipList-chipWrapper-1"
  >
    <div
      class="MuiChip-root ChipList-chip-2 MuiChip-colorSecondary"
    >
      <span
        class="MuiChip-label"
      >
        <p
          class="MuiTypography-root MuiTypography-body2"
        >
          Black Friendly
        </p>
      </span>
    </div>
    <div
      class="MuiChip-root ChipList-chip-2 MuiChip-colorSecondary"
    >
      <span
        class="MuiChip-label"
      >
        <p
          class="MuiTypography-root MuiTypography-body2"
        >
          Inclusive
        </p>
      </span>
    </div>
    <div
      class="MuiChip-root ChipList-chip-2 MuiChip-colorSecondary"
    >
      <span
        class="MuiChip-label"
      >
        <p
          class="MuiTypography-root MuiTypography-body2"
        >
          Black Owned
        </p>
      </span>
    </div>
    <div
      class="MuiChip-root ChipList-chip-2 MuiChip-colorSecondary"
    >
      <span
        class="MuiChip-label"
      >
        <p
          class="MuiTypography-root MuiTypography-body2"
        >
          Gender Neutral Restrooms
        </p>
      </span>
    </div>
    <div
      class="MuiChip-root ChipList-chip-2 MuiChip-colorSecondary"
    >
      <span
        class="MuiChip-label"
      >
        <p
          class="MuiTypography-root MuiTypography-body2"
        >
          Accessible
        </p>
      </span>
    </div>
    <div
      class="MuiChip-root ChipList-chip-2 MuiChip-colorSecondary"
    >
      <span
        class="MuiChip-label"
      >
        <p
          class="MuiTypography-root MuiTypography-body2"
        >
          Queer hangout space
        </p>
      </span>
    </div>
    <div
      class="MuiChip-root ChipList-chip-2 MuiChip-colorSecondary"
    >
      <span
        class="MuiChip-label"
      >
        <p
          class="MuiTypography-root MuiTypography-body2"
        >
          Trans friendly
        </p>
      </span>
    </div>
    <div
      class="MuiChip-root ChipList-chip-2 MuiChip-colorSecondary"
    >
      <span
        class="MuiChip-label"
      >
        <p
          class="MuiTypography-root MuiTypography-body2"
        >
          Queer owned
        </p>
      </span>
    </div>
  </div>
</DocumentFragment>
`;
